src/Proxy.NET/Program.cs

```diff
@@ -0,0 +1,135 @@
+using System.Net;
+using System.Text.Json;
+using Azure.Core;
+using Azure.Identity;
+using AzureOpenAIProxy.Management;
+using AzureOpenAIProxy.Management.Database;
+using Microsoft.EntityFrameworkCore;
+using Npgsql;
+using Proxy.NET.Models;
+using Proxy.NET.Routes;
+using Proxy.NET.Services;
+
+WebApplicationBuilder builder = WebApplication.CreateBuilder(args);
+
+NpgsqlDataSourceBuilder dataSourceBuilder = new(builder.Configuration.GetConnectionString("AoaiProxyContext"));
+dataSourceBuilder.MapEnum<ModelType>();
+
+if (string.IsNullOrEmpty(dataSourceBuilder.ConnectionStringBuilder.Password))
+{
+    // If no password was provided, we're probably going to be using a managed identity to authenticate.
+    // Define the time period the token lives for, and how long to wait to retry if the token fails.
+    int refreshPeriod = builder.Configuration.GetValue("Azure:TokenRefreshSeconds", 14400);
+    int failureRefreshPeriod = builder.Configuration.GetValue("Azure:TokenFailureRefreshSeconds", 10);
+
+    // Use the Azure SDK to get a token from the managed identity and provide it to Npgsql using the
+    // PeriodicPasswordProvider to refresh it on schedule.
+    dataSourceBuilder.UsePeriodicPasswordProvider(
+        async (_, ct) =>
+        {
+            var credential = new DefaultAzureCredential();
+            var ctx = new TokenRequestContext(["https://ossrdbms-aad.database.windows.net/.default"]);
+            var tokenResponse = await credential.GetTokenAsync(ctx, ct);
+            return tokenResponse.Token;
+        },
+        TimeSpan.FromSeconds(refreshPeriod),
+        TimeSpan.FromSeconds(failureRefreshPeriod)
+    );
+}
+
+NpgsqlDataSource dataSource = dataSourceBuilder.Build();
+
+builder.Services.AddDbContext<AoaiProxyContext>(
+    (option) =>
+    {
+        option.UseNpgsql(
+            dataSource,
+            // https://stackoverflow.com/questions/70423137/how-to-gracefully-handle-a-postgres-restart-in-npgsql
+            (options) =>
+            {
+                options.EnableRetryOnFailure(4, TimeSpan.FromSeconds(30), ["57P01"]);
+            }
+        );
+    }
+);
+
+builder.Services.AddAuthorization();
+builder.Services.AddMemoryCache();
+builder.Services.AddHttpClient<IProxyService, ProxyService>();
+builder.Services.AddProxyServices();
+builder.Services.AddApplicationInsightsTelemetry();
+
+var app = builder.Build();
+
+app.UseHttpsRedirection();
+app.UseAuthorization();
+app.MapProxyRoutes();
+
+// validate the request and authorize the user.
+// calling the next middleware if successful.
+// global catch exceptions and log them to App Insights
+app.Use(
```

Comment:
I really want us to revisit this. We shouldn't be writing our own middleware like this for auth, we should leverage the aspnet auth pipeline. Yes, we'd have to write our own auth provider, but then it plugs in properly.

Also, the global try/catch is not a good practice, we should have the route handlers return the right response types rather than throwing exceptions.

Reply:
done

Reply:
Can we remove this now?

Reply:
I think we need for the 429 for now. Also, you removed the OpenAI Exception formatting - I'm not sure that'll now work correctlt with the SDKs